### PR TITLE
PSI/XmlDocs: support resolve in C#

### PR DESCRIPTION
**Pull request by DedSec256**
image

### CODE REVIEW

**Comment by DedSec256** [comment]
we need to take the documentation from the signature file if it exists

File: ReSharper.FSharp/test/data/cache/csharpResolve/tooltips/XmlDoc 13 - Type members 02 - Abstract member.fsi [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module Test
+
+[<AbstractClass>]
+type A =
+    /// M signature
+    abstract member M: unit -> unit
+    /// M signature default
```
Comment (auduchinok) [inline ReSharper.FSharp/test/data/cache/csharpResolve/tooltips/XmlDoc 13 - Type members 02 - Abstract member.fsi:7]: It's a very interesting case, nice find. We should probably emit some warning here?

File: ReSharper.FSharp/test/data/cache/csharpResolve/tooltips/XmlDoc 13 - Type members 02 - Abstract member.fs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+module Test
+
+[<AbstractClass>]
+type A =
+    /// M 
+    abstract member M: unit -> unit
+    /// M default
```
Comment (auduchinok) [inline ReSharper.FSharp/test/data/cache/csharpResolve/tooltips/XmlDoc 13 - Type members 02 - Abstract member.fs:7]: And in other cases like this. 🙂

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/Tree/TopReferencePat.cs [hunk 3]
```diff
@@ -18,5 +18,7 @@ internal partial class TopReferencePat
     public TreeNodeCollection<IAttribute> Attributes =>
       this.GetBindingFromHeadPattern()?.Attributes ??
       TreeNodeCollection<IAttribute>.Empty;
+
+    public override XmlDocBlock XmlDocBlock => Parent?.FirstChild as XmlDocBlock;
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/Tree/TopReferencePat.cs:22]: Likely not a very common issue here, but something like this won't work:
```fsharp
/// X
let (x) = 1
```
Maybe use typed APIs like `BindingNavigator.GetByHeadPattern(this.IgnoreParentParens())` to get the declaration instead?

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/Tree/TopBinding.cs [hunk 4]
```diff
@@ -37,7 +37,9 @@ public TreeTextRange GetNameRange()
 
     public ITypeDeclaration GetContainingTypeDeclaration() => GetContainingNode<ITypeDeclaration>();
 
-    public XmlNode GetXMLDoc(bool inherit) => null;
+    public XmlNode GetXMLDoc(bool inherit) =>
+      FirstChild is XmlDocBlock xmlDocBlock ? xmlDocBlock.GetXML(null) : null;
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/Tree/TopBinding.cs:41]: Do we ever get the xml from the binding node and not from its head pattern declared element? If not, we could have an assertion here instead.

**Review by auduchinok — COMMENTED** [review]
@DedSec256 Very good testing! I've left two nits about making things saver and assertive, could you take a look?

File: ReSharper.FSharp/test/data/cache/csharpResolve/tooltips/XmlDoc 13 - Type members 02 - Abstract member.fsi (hunk 1 above)
Comment (DedSec256) [reply ReSharper.FSharp/test/data/cache/csharpResolve/tooltips/XmlDoc 13 - Type members 02 - Abstract member.fsi:7]: yes, sure

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/Tree/TopReferencePat.cs [hunk 5]
```diff
@@ -19,6 +19,7 @@ internal partial class TopReferencePat
       this.GetBindingFromHeadPattern()?.Attributes ??
       TreeNodeCollection<IAttribute>.Empty;
 
-    public override XmlDocBlock XmlDocBlock => Parent?.FirstChild as XmlDocBlock;
+    public override XmlDocBlock XmlDocBlock =>
+      this.IgnoreParentParens().GetBindingFromHeadPattern()?.FirstChild as XmlDocBlock;
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/Tree/TopReferencePat.cs:23]: What is `IgnoreParentParens` needed for here?

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/Tree/TopBinding.cs [hunk 6]
```diff
@@ -37,8 +38,11 @@ public TreeTextRange GetNameRange()
 
     public ITypeDeclaration GetContainingTypeDeclaration() => GetContainingNode<ITypeDeclaration>();
 
-    public XmlNode GetXMLDoc(bool inherit) =>
-      FirstChild is XmlDocBlock xmlDocBlock ? xmlDocBlock.GetXML(null) : null;
+    public XmlNode GetXMLDoc(bool inherit)
+    {
+      Assertion.Fail("Unexpected call TopBinding.GetXMLDoc");
+      return FirstChild is XmlDocBlock xmlDocBlock ? xmlDocBlock.GetXML(null) : null;
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/Tree/TopBinding.cs:44]: Could you simplify this dead code?

**Review by auduchinok — APPROVED** [review]
@DedSec256 Thanks, it's very nice to finally integrate better with R# here!

**Pull request merged**